src/eos-shard-alias-table.c

```diff
@@ -0,0 +1,138 @@
+
+#include "eos-shard-alias-table.h"
+
+#define FILTER_RECORD_NAME "d307a0144b2a9bfc5560eb1f7db921f0ca939e0c"
+#define JLIST_RECORD_NAME "c307a0144b2a9bfc5560eb1f7db921f0ca939e0c"
+#define MIMETYPE "application/octet-stream"
+#define FLAGS EOS_SHARD_BLOB_FLAG_NONE
+
+char *
+eos_shard_alias_table_find_entry (EosShardAliasTable *self, char *key)
+{
+  if (eos_shard_bloom_filter_test (self->filter, key)) {
+    return eos_shard_jlist_lookup_key (self->jlist, key);
+  }
+  return NULL;
+}
+
+/**
+ * eos_shard_alias_table_find_entries:
+ * @keys: (array length=n_keys):
+ *
+ * Returns: (transfer container) (element-type utf8 utf8): table
+ */
+GHashTable *
+eos_shard_alias_table_find_entries (EosShardAliasTable *self, char **keys, int n_keys)
+{
+  GHashTable *table = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  int i;
+  for (i=0; i<n_keys; i++) 
+    {
+      char *value = eos_shard_alias_table_find_entry (self, keys[i]);
+      g_hash_table_insert (table, g_strdup(keys[i]), value);
+    }
+  return table;
+}
+
+void
+eos_shard_alias_table_add_entry (EosShardAliasTable *self, char *key, char *value)
+{
+  eos_shard_jlist_writer_add_entry (self->jlist_writer, key, value);
+  eos_shard_bloom_filter_add (self->filter, key);
+}
+
+void
+eos_shard_alias_table_write_to_shard (EosShardAliasTable *self, EosShardWriter *writer)
+{
+  int filter_record_index, jlist_record_index;
+
+  eos_shard_jlist_writer_finish (self->jlist_writer);
+  eos_shard_bloom_filter_write_to_stream (self->filter, self->filter_out);
+
+  filter_record_index = eos_shard_writer_insert_record (writer, FILTER_RECORD_NAME);
+  eos_shard_writer_insert_blob (writer, EOS_SHARD_WRITER_BLOB_DATA, self->filter_file, MIMETYPE, FLAGS, filter_record_index);
+  eos_shard_writer_insert_blob (writer, EOS_SHARD_WRITER_BLOB_METADATA, self->filter_file, MIMETYPE, FLAGS, filter_record_index);
```

Comment:
This is a bizarre API. I think it should be more stateful:

```
eos_shard_writer_insert_record (writer, FILTER_RECORD_NAME);
eos_shard_writer_add_blob (writer, DATA, file, ...);
eos_shard_writer_add_blob (writer, METADATA, file, ...);
```

So add_blob pays attention to the last thing inserted, rather than the last record.